spectrum-offchain-cardano/src/data/cfmm_pool.rs

```diff
@@ -836,15 +1059,327 @@ impl ApplyOrder<ClassicalOnChainRedeem> for ConstFnPool {
 
-                Ok((self, redeem_output))
+                Ok((
+                    self,
+                    OperationResultBlueprint::single_output(redeem_output, validator),
+                ))
             }
             None => Err(ApplyOrderError::incompatible(redeem)),
         }
     }
 }
 
+impl<Ctx> ApplyOrder<OnChainRoyaltyWithdraw, Ctx> for ConstFnPool
+where
+    Ctx: Has<DeployedValidator<{ RoyaltyPoolRoyaltyWithdraw as u8 }>>
+        + Has<DeployedValidator<{ RoyaltyPoolV1RoyaltyWithdrawRequest as u8 }>>
+        + Has<RoyaltyWithdrawContext>,
+{
+    type Result = RoyaltyWithdrawOutput;
+
+    fn apply_order(
+        mut self,
+        royalty_withdraw: OnChainRoyaltyWithdraw,
+        ctx: Ctx,
+    ) -> Result<
+        (Self, OperationResultBlueprint<RoyaltyWithdrawOutput>),
+        ApplyOrderError<OnChainRoyaltyWithdraw>,
+    > {
+        let order = royalty_withdraw.order.clone();
+        let order_validator = royalty_withdraw.get_validator(&ctx);
+        let pool_order_validator: DeployedValidator<{ RoyaltyPoolRoyaltyWithdraw as u8 }> = ctx.get();
+        let royalty_withdraw_context: RoyaltyWithdrawContext = ctx.get();
+
+        let data_to_sign = DataToSign {
+            withdraw_data: royalty_withdraw.clone().order.into(),
+            pool_nonce: self.nonce,
+        };
+
+        let mut message_to_sign = order.additional_bytes.clone();
+        message_to_sign.extend(&data_to_sign.into_pd().to_cbor_bytes());
+
+        let public_key: PublicKey = self.royalty_pub_key.try_into().unwrap();
+
+        let signature_is_correct = public_key.verify(&message_to_sign, &order.signature);
+
+        if !signature_is_correct {
+            return Err(ApplyOrderError::verification_failed(
+                royalty_withdraw,
+                format!("signature_is_correct: {}", signature_is_correct),
+            ));
+        }
+
+        self.royalty_x = self
+            .royalty_x
+            .checked_sub(&order.withdraw_royalty_x)
+            .ok_or(ApplyOrderError::incompatible(royalty_withdraw.clone()))?;
+
+        self.royalty_y = self
+            .royalty_y
+            .checked_sub(&order.withdraw_royalty_y)
+            .ok_or(ApplyOrderError::incompatible(royalty_withdraw.clone()))?;
+
+        self.reserves_x = self
+            .reserves_x
+            .checked_sub(&order.withdraw_royalty_x)
+            .ok_or(ApplyOrderError::incompatible(royalty_withdraw.clone()))?;
+
+        self.reserves_y = self
+            .reserves_y
+            .checked_sub(&order.withdraw_royalty_y)
+            .ok_or(ApplyOrderError::incompatible(royalty_withdraw.clone()))?;
+
+        self.nonce += 1;
+
+        let royalty_output = RoyaltyWithdrawOutput {
+            token_x_asset: self.asset_x,
+            token_x_amount: order.withdraw_royalty_x,
+            token_y_asset: self.asset_y,
+            token_y_amount: order.withdraw_royalty_y,
+            ada_residue: order.init_ada_value - order.fee,
```

Comment:
No overflow, sure?

Reply:
Same as for pool, add verification at parsing stage